{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dec0367a-eab7-42cc-bb34-c5aad23da69b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import re\n",
    "import argparse\n",
    "import pysam\n",
    "from pyfaidx import Fasta\n",
    "from Bio.Seq import Seq\n",
    "from Bio import pairwise2"
   ]
  },
  {
   "cell_type": "raw",
   "id": "81d5d276-a995-408f-9019-a52b51ffb271",
   "metadata": {},
   "source": [
    "'''\n",
    "From first read (query_sequence) in B1_EKDL230008374_R12_Araport_Bait_R12.sorted.bam (note the lower case 't' in the ref_sequence):                                                                                                                                                   \n",
    "ref_sequence  (SmB-A_AT5G44500.1)                                     CATCGACATCTTCAGACTACTAGTTTGATTCGACCCACCGATACCGTTCGTATAATGTATGCTATACGAACGGTATCACAAGTTTGtACAAAAAAGTTGG\n",
    "aso.query_sequence[aso.query_alignment_start:aso.query_alignment_end] CATCGACATCTTCAGACTACTAGTTTGATTCGACCCACCGATACCGTTCGTATAATGTATGCTATACGAACGGTATCACAAGTTTGAACAAAAAAGTTGG\n",
    "query_sequence                TCCTGTAGTTTATGAACTGTAGCATCTTGGAGCTCTTCGACATCGACATCTTCAGACTACTAGTTTGATTCGACCCACCGATACCGTTCGTATAATGTATGCTATACGAACGGTATCACAAGTTTGAACAAAAAAGTTGGACTTTATCTCCCAAAACACAAAACAAAAAAAATGGCTTCCACTGCTCTCTCAAGCGCC\n",
    "aso.query_sequence[140:] (Will match AT1G60950.1)                                                                                                                         ACTTTATCTCCCAAAACACAAAACAAAAAAAATGGCTTCCACTGCTCTCTCAAGCGCC\n",
    "\n",
    "\n",
    "SmB-A_AT5G44500.1 from excel file (concatenated columns):\n",
    "\n",
    "CATCGACATCTTCAGACTACTAGTTTGATTCGACCCACCGATACCGTTCGTATAATGTATGCTATACGAACGGTATCACAAGTTTGTACAAAAAAGTTGG\n",
    "||||||||| (\"gene\" no mismatches)\n",
    "CATCGACAT\n",
    "         |||||| (common)\n",
    "         CTTCAG\n",
    "               ||| (ACT or GGA)\n",
    "               ACT\n",
    "                  ||||||||||||||||| (common)\n",
    "                  ACTAGTTTGATTCGACC\n",
    "                                   |||| (common)\n",
    "                                   CACC\n",
    "                                       || (common)\n",
    "                                       GA\n",
    "                                         |||||||||||||||||||||||||||||||||| (allow 2 mismatches)\n",
    "                                         TACCGTTCGTATAATGTATGCTATACGAACGGTA\n",
    "                                                                           ||||||||||||||||||||||||| (allow 2 mismatches)\n",
    "                                                                           TCACAAGTTTGTACAAAAAAGTTGG\n",
    "\n",
    "\n",
    "\n",
    "ref_sequence (printed from first read in bam):\n",
    "CATCGACATCTTCAGACTACTAGTTTGATTCGACCCACCGATACCGTTCGTATAATGTATGCTATACGAACGGTATCACAAGTTTGtACAAAAAAGTTGG\n",
    "\n",
    "\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d0b75cb2-dcbb-4699-9e73-2ebb8418f5ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Tmp file input\n",
    "bam_file = 'tmp/B1_EKDL230008374.merged.bam'\n",
    "baits_file = 'tmp/Bait_Gateway_AtcDNAlibrary.fas'\n",
    "araport_file = 'tmp/Araport11_cdna_20220914_representative_gene_model.fas'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "be046f5a-0e5a-42fa-840c-f836b43f0726",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Defaults\n",
    "__version__ = '0.1'\n",
    "DEBUG = 0\n",
    "mismatches_default = 2\n",
    "allowed_hd = mismatches_default # tmp\n",
    "threads_default = 4\n",
    "\n",
    "## Coordinates on bait sequences\n",
    "min_length_seq_bait_part = 100\n",
    "gene_length_default = 9\n",
    "gene_length = gene_length_default # tmp\n",
    "accepted_index1 = ['ACT', 'GGA']\n",
    "accepted_eco571 = ['CTTCAG']\n",
    "accepted_linker1 = ['ACTAGTTTGATTCGACC']\n",
    "accepted_index2 = ['CACC']\n",
    "accepted_linker2 = ['GA']\n",
    "\n",
    "### Pos 1-9 is the \"gene\": CATCGACAT, ...\n",
    "gene_start = 0\n",
    "gene_end = gene_length\n",
    "### Pos 10-15 is Eco57I: CTTCAG\n",
    "eco571_start = 9\n",
    "eco571_end = 15\n",
    "### Pos 16-18 is the index: ACT or GGA\n",
    "index1_start = 15\n",
    "index1_end = 18\n",
    "### Pos 19-35 is a linker: ACTAGTTTGATTCGACC\n",
    "linker1_start = 18\n",
    "linker1_end = 35\n",
    "### Pos 36-39 is an index: CACC\n",
    "index2_start = 35\n",
    "index2_end = 39\n",
    "### Pos 40-41 is a linker: GA\n",
    "linker2_start = 39\n",
    "linker2_end = 41\n",
    "### Pos 42-75 is lox66/71: TACCGTTCGTATAATGTATGCTATACGAACGGTA\n",
    "lox6671_start = 41\n",
    "lox6671_end = 75\n",
    "### Pos 76-100 is \"after Gateway recombination\": TCACAAGTTTGTACAAAAAAGTTGG\n",
    "agr_start = 75\n",
    "agr_end = 100"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e2ff0a0f-9c8b-4a3d-88c2-a9d9843c4f3d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def hammingDist(seq1, seq2):\n",
    "    \"\"\"\n",
    "    Calculate the Hamming distance in terms of number of changes\n",
    "    Args: seq1 and seq2 (str)\n",
    "    Return: distance (int)\n",
    "    Note: we use .upper() to always compare upper case\n",
    "    \"\"\"\n",
    "    i = 0\n",
    "    count = 0\n",
    "    while (i < len(seq1)):\n",
    "        if (seq1[i].upper() != seq2[i].upper()):\n",
    "            count += 1\n",
    "        i += 1\n",
    "    return count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e4b2d854-fb53-4983-b33f-1feaca1688b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Read bam file. Needs to be indexed: `samtools index -@ 8 B1_EKDL230008374.merged.bam`\n",
    "bam = pysam.AlignmentFile(bam_file, 'rb', threads=2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0e070b39-b93e-41fc-8d24-05030994f611",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Read baits file\n",
    "baits = Fasta(baits_file, sequence_always_upper=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5aaa69ba-cfa3-4d72-abc0-f8615ef7cf74",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Read baits_file and create a hashes with key=SmB-A_AT5G44500.1, value=CATCGACAT (\"gene\") and value=gene_count\n",
    "baits_gene_seq_dict = {}\n",
    "baits_gene_name_dict = {}\n",
    "baits_gene_count_dict = {}\n",
    "\n",
    "for seq in baits:\n",
    "    baits_gene_seq_dict[seq.name] = seq[0:gene_length]\n",
    "    baits_gene_name_dict[str(seq[0:gene_length])] = seq.name\n",
    "    baits_gene_count_dict[seq.name] = 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b22d01a3-40ec-4771-823f-032aa7a10574",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Read araport file\n",
    "araport = Fasta(araport_file, sequence_always_upper=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8bf320ce-d13a-4566-8d42-bb4bee9d047d",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Tmp counters\n",
    "len_seq_bait_part_ok_counter = 0\n",
    "len_seq_bait_part_not_ok_counter = 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3f204640-537c-46ab-b735-919258855d4c",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "## Test loop\n",
    "for aso in bam.head(1):\n",
    "    #help(aso)\n",
    "    #print(\"query name \", aso.query_name)\n",
    "    #print(\"reference name \", aso.reference_name)\n",
    "    #print(\"reference sequence \", aso.get_reference_sequence())\n",
    "    #print(baits[aso.reference_name][41:75])\n",
    "    \n",
    "    ## Araport part\n",
    "    start_araport_part = aso.query_alignment_end\n",
    "    seq_araport_part = Seq(str(aso.query_sequence[start_araport_part:]))\n",
    "    #print(seq_araport_part) # ACTTTATCTCCCAAAACACAAAACAAAAAAAATGGCTTCCACTGCTCTCTCAAGCGCC\n",
    "    #query_name = aso.query_name.split(':')[0]   # AT1G60950.1\n",
    "    #print(query_name)\n",
    "    #print(str(araport[query_name])) # 985 bp sequence\n",
    "    #seq_araport_db = Seq(str(araport[query_name]))\n",
    "    #alignments = pairwise2.align.localms(seq_araport_part, seq_araport_db, match=2, mismatch=-1, open=-5, extend=-1, one_alignment_only=True)\n",
    "    #for match in alignments:\n",
    "    #    print(match)\n",
    "    ## Translation\n",
    "    #trsl = seq_araport_part.translate(table=11, cds=True) # Error\n",
    "    #fixed_seq = Seq('ACTTTATCTCCCAAAACACAAAACAAAAAAAA-TGGCTTCCACTGCTCTCTCAAGCGCC')\n",
    "    #fixed_seq.translate(table=11, cds=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "93aa77fd-212e-493b-af3d-ca9253cca5b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Big loop\n",
    "\n",
    "## Get aligned sequence object\n",
    "#for aso in bam.head(100):\n",
    "for aso in bam.fetch():\n",
    "\n",
    "    ## Bait part\n",
    "    seq_bait_part = Seq(str(aso.query_sequence[aso.query_alignment_start:aso.query_alignment_end]))\n",
    "\n",
    "    ## Discard read if length of seq_bait_part is lower than min_length_seq_bait_part (100)\n",
    "    if len(seq_bait_part) < min_length_seq_bait_part:\n",
    "        if DEBUG:\n",
    "            print(f\"DEBUG: SKIPPING: Length of seq_bait_part was lower than min_length_seq_bait_part ({len(seq_bait_part)})\")\n",
    "        len_seq_bait_part_not_ok_counter = len_seq_bait_part_not_ok_counter + 1\n",
    "        continue\n",
    "\n",
    "    ## Tmp counter\n",
    "    len_seq_bait_part_ok_counter = len_seq_bait_part_ok_counter + 1\n",
    "\n",
    "    ## Go through the parts by position and check if OK\n",
    "    ## Pos 1-9 is the \"gene\": CATCGACAT, ...\n",
    "    gene_seq = seq_bait_part[gene_start:gene_end]\n",
    "    if baits_gene_name_dict.get(gene_seq) is None:\n",
    "        if DEBUG:\n",
    "            print(f\"DEBUG: SKIPPING: did NOT find gene {gene_seq} in dict\")\n",
    "        continue\n",
    "    if DEBUG > 1:\n",
    "        print(f\"DEBUG: found gene in dict: {gene_seq}\")\n",
    "\n",
    "    ## Pos 10-15 is Eco57I: CTTCAG\n",
    "    if seq_bait_part[eco571_start:eco571_end] not in accepted_eco571:\n",
    "        if DEBUG:\n",
    "            print(f\"DEBUG: SKIPPING: Eco57I NOT accepted: {seq_bait_part[eco571_start:eco571_end]}\")\n",
    "        continue\n",
    "    if DEBUG > 1:\n",
    "        print(f\"DEBUG:  found Eco57I: {seq_bait_part[eco571_start:eco571_end]}\")\n",
    "\n",
    "    ## Pos 16-18 is the index1: ACT or GGA\n",
    "    if seq_bait_part[index1_start:index1_end] not in accepted_index1:\n",
    "        if DEBUG:\n",
    "            print(f\"DEBUG: SKIPPING: index1 NOT accepted: {seq_bait_part[index1_start:index1_end]}\")\n",
    "        continue\n",
    "    if DEBUG > 1:\n",
    "        print(f\"DEBUG:   found index1: {seq_bait_part[index1_start:index1_end]}\")\n",
    "\n",
    "    ## Pos 19-35 is a linker1: ACTAGTTTGATTCGACC\n",
    "    if DEBUG > 1:\n",
    "        print(f\"DEBUG:    found linker1: {seq_bait_part[linker1_start:linker1_end]}\")\n",
    "\n",
    "    ## Pos 36-39 is an index2: CACC\n",
    "    if DEBUG > 1:\n",
    "        print(f\"DEBUG:     found index2: {seq_bait_part[index2_start:index2_end]}\")\n",
    "\n",
    "    ## Pos 40-41 is a linker2: GA\n",
    "    if DEBUG > 1:\n",
    "        print(f\"DEBUG:      found linker2: {seq_bait_part[linker2_start:linker2_end]}\")\n",
    "\n",
    "    ## Pos 42-75 is lox66/71: TACCGTTCGTATAATGTATGCTATACGAACGGTA\n",
    "    read_lox_seq = seq_bait_part[lox6671_start:lox6671_end]\n",
    "    ref_lox_seq = baits[aso.reference_name][lox6671_start:lox6671_end]\n",
    "    hd = hammingDist(str(read_lox_seq), str(ref_lox_seq))\n",
    "    if hd <= allowed_hd:\n",
    "        if DEBUG > 1:\n",
    "            print(f\"DEBUG:       found lox66/71: {seq_bait_part[lox6671_start:lox6671_end]}\")\n",
    "    else:\n",
    "        if DEBUG:\n",
    "            print(f\"DEBUG: SKIPPING: HD giher than {allowed_hd} for lox66/71: {seq_bait_part[lox6671_start:lox6671_end]}\")\n",
    "        continue\n",
    "\n",
    "    ## Pos 76-100 is \"after Gateway recombination\": TCACAAGTTTGTACAAAAAAGTTGG\n",
    "    read_agr_seq = seq_bait_part[agr_start:agr_end]\n",
    "    ref_agr_seq = baits[aso.reference_name][agr_start:agr_end]\n",
    "    hd = hammingDist(str(read_agr_seq), str(ref_agr_seq))\n",
    "    if hd <= allowed_hd:\n",
    "        if DEBUG > 1:\n",
    "            print(f\"DEBUG:        found after Gateway recombination: {seq_bait_part[agr_start:agr_end]}\")\n",
    "    else:\n",
    "        if DEBUG:\n",
    "            print(f\"DEBUG: SKIPPING: HD giher than {allowed_hd} for after Gateway recombination: {seq_bait_part[agr_start:agr_end]}\")\n",
    "        continue\n",
    "\n",
    "    ## If All parts of seq_bait_part are OK, parse Araport part\n",
    "    ## Araport part\n",
    "    #start_araport_part = aso.query_alignment_end\n",
    "    #seq_araport_part = Seq(str(aso.query_sequence[start_araport_part:]))\n",
    "\n",
    "    #query_name = aso.query_name.split(':')[0]   # AT1G60950.1\n",
    "    #seq_araport_db = Seq(str(araport[query_name]))\n",
    "    \n",
    "    ##   Check reading frame for seq_araport_part\n",
    "    #alignments = pairwise2.align.localms(seq_araport_part, seq_araport_db, match=2, mismatch=-1, open=-5, extend=-1, one_alignment_only=True)\n",
    "    #for match in alignments:\n",
    "    #    #print(match)\n",
    "    #    print(match.seqA)  # seq_araport_part\n",
    "    #    print(match.start) # Start pos on seq_araport_db\n",
    "    #    print(match.end)   # End pos on seq_araport_db\n",
    "    \n",
    "    ##   Add info on in frame/out of frame\n",
    "\n",
    "    ## Tmp:  Add to count gene_count\n",
    "    baits_gene_count_dict[aso.reference_name] = baits_gene_count_dict[aso.reference_name] + 1\n",
    "    \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2b8a55fd-af2c-41c7-b3b2-999893ccdf81",
   "metadata": {},
   "outputs": [],
   "source": [
    "# End\n",
    "bam.close()\n",
    "\n",
    "print(f\"Bam file: {bam_file}\")\n",
    "print(f\"Nr bait length OK: {len_seq_bait_part_ok_counter}\")\n",
    "print(f\"Nr bait length not OK: {len_seq_bait_part_not_ok_counter}\")\n",
    "print(f\"Counts:\")\n",
    "\n",
    "for k,v in sorted(baits_gene_count_dict.items(), key=lambda x:x[1], reverse=True):\n",
    "    print(k, \"=\", v)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
